[IMP] stock: Faster stock.move migration

yajo opened the pull request:
This patch solves a bug and bunch of bottlenecks that were slowing down the execution of `move_stock_qty()`. The summary of changes is:

1. Inside `quants_unreserve()` I replaced a generator call for the much faster `.ids`. I know the usual thing is to not touch Odoo code, but this change doesn't affect logic but just performance.
2. Use sets instead of tuples for the `in` operator, always faster.
3. Avoid a bug in `_move_assign()` and `_move_done()` where the same outer variable name was being reused inside a loop, changing the result in unclean ways.
4. Index some fields that will be searched *a lot*.
5. Disable automatic field prefetching. Migrated models have lots of computed and relational fields that end up consuming more time than doing 1 query per field when needed.
6. Prefetch needed fields of as many records as possible before running the code, to avoid the previous point affecting the process negatively.
7. Add more logging so you can know the progress. This can be a big logging black hole otherwise.

Before this patch, it was processing about 3 moves/s in my server. After, it took about 10 mins prefetching things, but then it processed an average of 12 moves/s with peaks of more than 30 moves/s. That's 4x faster!

@Tecnativa TT18838

pedrobaeza commented on addons/stock/stock.py:
```diff
@@ -612,7 +612,7 @@ def _price_update(self, cr, uid, ids, newprice, context=None):
 
     def quants_unreserve(self, cr, uid, move, context=None):
-        related_quants = [x.id for x in move.reserved_quant_ids]
+        related_quants = move.reserved_quant_ids.ids
```
Is this so critical that the change is justified?

yajo replied:
Yes, this saves a lot of time.

pedrobaeza commented on addons/stock/migrations/8.0.1.1/post-migration.py:
```diff
@@ -926,16 +927,85 @@ def migrate_stock_qty(cr, registry):
+            'split_from',
+            'state',
+        ])
+        locations = moves.mapped("location_id").read([
```
Why are you storing the `read` instead of the recordset but on `products` is the other way?

yajo replied:
From here I only need the count, to log it.

pedrobaeza replied:
Yeah, but this is more memory than only storing the recordsets.

pedrobaeza commented on addons/stock/migrations/8.0.1.1/post-migration.py:
```diff
@@ -926,16 +927,85 @@ def migrate_stock_qty(cr, registry):
+            'type',
+            'uom_id',
+        ])
+        uoms = products.mapped("uom_id").read([
```
Same doubt than locations

yajo replied:
From here I only need the count, to log it.

yajo committed:
[IMP] stock: Faster stock.move migration

This patch solves a bug and bunch of bottlenecks that were slowing down the execution of `move_stock_qty()`. The summary of changes is:

1. Inside `quants_unreserve()` I replaced a generator call for the much faster `.ids`. I know the usual thing is to not touch Odoo code, but this change doesn't affect logic but just performance.
2. Use sets instead of tuples for the `in` operator, always faster.
3. Avoid a bug in `_move_assign()` and `_move_done()` where the same outer variable name was being reused inside a loop, changing the result in unclean ways.
4. Index some fields that will be searched *a lot*.
5. Disable automatic field prefetching. Migrated models have lots of computed and relational fields that end up consuming more time than doing 1 query per field when needed.
6. Prefetch needed fields of as many records as possible before running the code, to avoid the previous point affecting the process negatively.
7. Add more logging so you can know the progress. This can be a big logging black hole otherwise.

Before this patch, it was processing about 3 moves/s in my server. After, it took about 10 mins prefetching things, but then it processed an average of 12 moves/s with peaks of more than 30 moves/s. That's 4x faster!

@Tecnativa TT18838

yajo commented:
I'm not sure that's a problem, because the purpose of prefetching is
keeping things in memory.

In any case, I changed it, just in case.

El vie., 8 nov. 2019 a las 10:54, Pedro M. Baeza (<notifications@github.com>)
escribió:

> *@pedrobaeza* commented on this pull request.
> ------------------------------
>
> In addons/stock/migrations/8.0.1.1/post-migration.py
> <https://github.com/OCA/OpenUpgrade/pull/2080#discussion_r344120149>:
>
> > +            'location_dest_id',
> +            'location_id',
> +            'move_orig_ids',
> +            'origin_returned_move_id',
> +            'partially_available',
> +            'picking_id',
> +            'product_id',
> +            'product_qty',
> +            'reserved_availability',
> +            'reserved_quant_ids',
> +            'restrict_lot_id',
> +            'restrict_partner_id',
> +            'split_from',
> +            'state',
> +        ])
> +        locations = moves.mapped("location_id").read([
>
> Yeah, but this is more memory than only storing the recordsets.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/OpenUpgrade/pull/2080?email_source=notifications&email_token=AAHNXDIOUFASYDG2ILLJPYDQSVAQHA5CNFSM4JKUGSIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCK4TWCY#discussion_r344120149>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHNXDL4QRNC6XH4MQAM6NLQSVAQHANCNFSM4JKUGSIA>
> .
>


-- 
Jairo Llopis

pedrobaeza commented:
I meant "Python memory", so thanks for changing it.
